src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp

```diff
@@ -64,11 +66,101 @@ struct ReplicatedMergeTreeSinkImpl<async_insert>::DelayedChunk
                 block_id_to_offset_idx.clear();
                 for (size_t i = 0; i < block_id.size(); ++i)
                 {
-                    block_id_to_offset_idx[block_id[i]] = i;
+                    block_id_to_offset_idx[block_id[i]].push_back(i);
                 }
             }
         }
 
+        bool checkSelfDeduplicate()
+        {
+            if constexpr (async_insert)
+            {
+                std::vector<String> dup_block_ids;
+                for (const auto & [hash_id, offset_idxes] : block_id_to_offset_idx)
+                {
+                    if (offset_idxes.size() > 1)
+                        dup_block_ids.push_back(hash_id);
+                }
+                if (dup_block_ids.empty())
+                    return false;
+
+                rewriteBlock(dup_block_ids, true);
+                return true;
+            }
+            return false;
+        }
+
+        /// remove the conflict parts of block for rewriting again.
+        void rewriteBlock(const std::vector<String> & block_paths, bool self_dedup)
```

Comment:
Actually we don't `write` block here. Maybe rename to something like `filterBlockDuplicates`?